### ISSUE TITLE
Error on emacs startup

### ISSUE THREAD

**Issue by stardiviner** [nicolaisingh/saveplace-pdf-view#9]
I got error on Emacs startup with `saveplace-pdf-view` enabled.

```
Debugger entered--Lisp error: (wrong-type-argument stringp ("Clojure for the Brave and True.pdf" (filename . "~/Org/Wiki/Computer Technology/Programming/Program...") (position . 1) (last-modified 25723 24847 287081 0) (page . 10) (slice 0.074206 0.129655 0.792064 0.7857350000000001) (size . fit-width) (origin 0.0 . 0.0) (handler . pdf-view-bookmark-jump-handler)))
  expand-file-name(("Clojure for the Brave and True.pdf" (filename . "~/Org/Wiki/Computer Technology/Programming/Program...") (position . 1) (last-modified 25723 24847 287081 0) (page . 10) (slice 0.074206 0.129655 0.792064 0.7857350000000001) (size . fit-width) (origin 0.0 . 0.0) (handler . pdf-view-bookmark-jump-handler)))
  #f(compiled-function (sym val) #<bytecode 0x1267272fdd97191>)(save-place-abbreviate-file-names nil)
  custom-initialize-reset(save-place-abbreviate-file-names (funcall #'#f(compiled-function () #<bytecode 0x19800016fe914>)))
  custom-declare-variable(save-place-abbreviate-file-names (funcall #'#f(compiled-function () #<bytecode 0x19800016fe914>)) "If non-nil, abbreviate file names before saving th..." :type boolean :set #f(compiled-function (sym val) #<bytecode 0x1267272fdd97191>) :version "28.1")
  byte-code("\300\301\302\303\304DD\305\306\307\310\311\312\313&\11\210\300\314\302\303\315DD\316\306\307%\210\300\317\302\303\320DD\321\306\322%\210\300\323\302\303\324DD\325\312..." [custom-declare-variable save-place-abbreviate-file-names funcall function #f(compiled-function () #<bytecode 0x19800016fe914>) "If non-nil, abbreviate file names before saving th..." :type boolean :set #f(compiled-function (sym val) #<bytecode 0x1267272fdd97191>) :version "28.1" save-place-save-skipped #f(compiled-function () #<bytecode 0x19800016fe5d4>) "If non-nil, remember files matching `save-place-sk..." save-place-skip-check-regexp #f(compiled-function () #<bytecode -0x7810fef85890ea4>) "Regexp whose file names shall not be checked for r..." regexp save-place-ignore-files-regexp #f(compiled-function () #<bytecode -0x6b951d2e7268727>) "Regexp matching files for which no position should..." "24.1"] 10)
  save-place-mode()
  run-hooks(after-init-hook delayed-warnings-hook)
  command-line()
  normal-top-level()
```

Here is my package config:
```elisp
(use-package saveplace-pdf-view
  :ensure t)
```

My Emacs version: GNU Emacs 30.0.50 (build 1, aarch64-apple-darwin22.5.0, NS appkit-2299.60 Version 13.4 (Build 22F66)) of 2023-06-02

Here is the part content of my saveplace session file:

```elisp
(...
("/Users/stardiviner/Org/Wiki/Computer Technology/Programming/Programming Languages/Python/Python Packages/Jupyter/Jupyter.org" . 85676)
 ("/Users/stardiviner/Org/Wiki/Computer Technology/Software/Softwares/JupyterLab.org" . 13665)
 ("/Users/stardiviner/Code/Emacs/chatgpt-shell/chatgpt-shell.el" . 50668)
 ("/Users/stardiviner/.config/emacs/elpa/shell-maker-20230524.2217/shell-maker.el" . 30770)
 ("/Users/stardiviner/Org/Wiki/Computer Technology/Programming/Programming Languages/Clojure/books/"
  (dired-filename . "/Users/stardiviner/Org/Wiki/Computer Technology/Programming/Programming Languages/Clojure/books/Clojure for the Brave and True.pdf"))
 ("/Users/stardiviner/Org/Wiki/Computer Technology/Programming/Programming Languages/Clojure/books/Clojure for the Brave and True.pdf"
  (pdf-view-bookmark "Clojure for the Brave and True.pdf"
                     (filename . "~/Org/Wiki/Computer Technology/Programming/Programming Languages/Clojure/books/Clojure for the Brave and True.pdf")
                     (position . 1)
                     (last-modified 25723 24847 287081 0)
                     (page . 10)
                     (slice 0.074206 0.129655 0.792064 0.7857350000000001)
                     (size . fit-width)
                     (origin 0.0 . 0.0)
                     (handler . pdf-view-bookmark-jump-handler)))
 ("/Users/stardiviner/.zshrc" . 12932)
 ("/Users/stardiviner/Org/Wiki/Computer Technology/Programming/Programming Languages/Python/Python Packages/Anaconda.org" . 32899)
 ("/Users/stardiviner/.config/emacs/init/Emacs/init-emacs-appearance.el" . 5864)
 ("/Users/stardiviner/Org/Wiki/Computer Technology/Programming/Emacs/Emacs Packages/spacious-padding.org" . 7841)
 ("/Users/stardiviner/.config/emacs/elpa/spacious-padding-0.0.0/spacious-padding.el" . 1808)
 ("/Users/stardiviner/.config/emacs/custom.el" . 2)
...
)
```

**Comment by stardiviner** [comment]
I doubt that the saveplace-pdf-view session info is not compatible with new version Emacs?

**Comment by nicolaisingh** [comment]
Hi, I haven't tested this package yet with the latest Emacs unfortunately.  Will check if I find the time!  Thanks for reporting this.

**Comment by overideal** [comment]
I can confirm that the package does not work with Emacs 30.0.50 (on Linux).

**Comment by 9viz** [comment]
I have submitted a patch to Emacs to hopefully resolve this issue.  See https://debbugs.gnu.org/cgi/bugreport.cgi?bug=65055

**Comment by 9viz** [comment]
The patch was understandably not applied.  But here's a diff to fix the issue on saveplace-pdf-view's side.

```
diff --git a/saveplace-pdf-view.el b/saveplace-pdf-view.el
index 05e7206..6ad5c0c 100644
--- a/saveplace-pdf-view.el
+++ b/saveplace-pdf-view.el
@@ -55,10 +55,10 @@ persisted bookmark under SAVE-PLACE-ALIST-KEY."
   (or save-place-loaded (load-save-place-alist-from-file))
   (let* ((cell (assoc buffer-file-name save-place-alist)))
     (when (and cell
-               (listp (cdr cell))
-               (assq save-place-alist-key (cdr cell)))
+               (vectorp (cdr cell))
+               (assq save-place-alist-key (aref (cdr cell) 0)))
       (funcall bookmark-jump-function
-               (cdr (assq save-place-alist-key (cdr cell)))))
+               (cdr (assq save-place-alist-key (aref (cdr cell) 0)))))
     (setq save-place-mode t)))
 
 (defun saveplace-pdf-view-to-alist (save-place-alist-key make-record-function)
@@ -93,7 +93,7 @@ doc-view bookmark will replace the file's pdf-view bookmark."
                                (or (null (cdr origin))
                                    (equal '(0.0 . 0.0) (cdr origin))))))
             (setq save-place-alist
-                  (cons (cons item `((,save-place-alist-key . ,bookmark)))
+                  (cons (cons item (vector `((,save-place-alist-key ,bookmark))))
                         save-place-alist))))))))
 
 (defun saveplace-pdf-view-find-file-advice (orig-fun &rest args)
```

**Comment by stardiviner** [comment]
Thanks @vizs 

Updated patch for latest commit "54ed966"

```diff
From cacfe72df5aa2a3b6ada9f4febbe1bf5986f14c1 Mon Sep 17 00:00:00 2001
From: stardiviner <numbchild@gmail.com>
Date: Sat, 5 Aug 2023 19:33:20 +0800
Subject: [PATCH] Fix #9 Emacs new version saveplace compatibility.

---
 saveplace-pdf-view.el | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/saveplace-pdf-view.el b/saveplace-pdf-view.el
index 05e7206..0ec65bf 100644
--- a/saveplace-pdf-view.el
+++ b/saveplace-pdf-view.el
@@ -55,10 +55,10 @@ persisted bookmark under SAVE-PLACE-ALIST-KEY."
   (or save-place-loaded (load-save-place-alist-from-file))
   (let* ((cell (assoc buffer-file-name save-place-alist)))
     (when (and cell
-               (listp (cdr cell))
-               (assq save-place-alist-key (cdr cell)))
+               (vectorp (cdr cell))
+               (assq save-place-alist-key (aref (cdr cell) 0)))
       (funcall bookmark-jump-function
-               (cdr (assq save-place-alist-key (cdr cell)))))
+               (cdr (assq save-place-alist-key (aref (cdr cell) 0)))))
     (setq save-place-mode t)))
 
 (defun saveplace-pdf-view-to-alist (save-place-alist-key make-record-function)
@@ -93,7 +93,7 @@ doc-view bookmark will replace the file's pdf-view bookmark."
                                (or (null (cdr origin))
                                    (equal '(0.0 . 0.0) (cdr origin))))))
             (setq save-place-alist
-                  (cons (cons item `((,save-place-alist-key . ,bookmark)))
+                  (cons (cons item (vector `((,save-place-alist-key . ,bookmark))))
                         save-place-alist))))))))
 
 (defun saveplace-pdf-view-find-file-advice (orig-fun &rest args)
-- 
2.39.2 (Apple Git-143)

```

**Comment by nicolaisingh** [comment]
It seems Emacs 30 introduced some changes in `saveplace` related to filename abbreviation and caused this package to break.  This fix works with both Emacs 29 and 30, in `pdf-view-mode` and also `doc-view-mode`.

Thanks @vizs @stardiviner ! :rocket:
